File: 404.sum-of-left-leaves/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sumOfLeftLeaves(root: TreeNode?) = sum(root, false)
    
    private fun sum(node: TreeNode?, isLeft: Boolean): Int = 
        if (node == null) {
            0
        } else {
            with (node) {
                if (left == null && right == null && isLeft) {
                    `val`
                } else {
                    sum(left, true) + sum(right, false)
                }
            }
        }
}

File: 404.sum-of-left-leaves/solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */
object Solution {
    def sumOfLeftLeaves(root: TreeNode): Int = {
        def sum(node: TreeNode, isLeft: Boolean): Int = {
            if (node == null) 0
            else if (isLeft && isLeaf(node)) node.value
            else sum(node.left, true) + sum(node.right, false)
        }
        def isLeaf(node: TreeNode) =
            node != null && node.left == null && node.right == null
        sum(root, false)
    }
}